RUFAS/input_manager.py

```diff
@@ -142,31 +159,171 @@ def _validate_data(self, eager_termination: bool = True) -> bool:
+
+    def _validate_array_type_element(self, variable_to_check: Dict[str, Any], var_name: str,
+                                     input_data_value: list) -> bool:
+        """Validates a __pool element of type array."""
+        info_map = {"class": self.__class__.__name__,
+                    "function": self._validate_array_type_element.__name__,
+                    }
+        is_maximum_length = variable_to_check.get("maximum_length")
+        is_minimum_length = variable_to_check.get("minimum_length")
+        if is_maximum_length and is_minimum_length:
+            is_in_range = variable_to_check["minimum_length"] <= len(input_data_value) <= \
+                variable_to_check["maximum_length"]
+            if not is_in_range:
+                om.add_warning("Array out of length range.", f"{var_name=}", info_map)
+                return is_in_range
+        if is_minimum_length:
+            is_in_range = variable_to_check["minimum_length"] <= len(input_data_value)
+            if not is_in_range:
+                om.add_warning("Array out of length range.", f"{var_name=}", info_map)
+                return is_in_range
+        if is_maximum_length:
+            is_in_range = len(input_data_value) <= variable_to_check["maximum_length"]
+            if not is_in_range:
+                om.add_warning("Array out of length range.", f"{var_name=}", info_map)
+                return is_in_range
 
-        # TODO in validate_element fun branch
-        # where element is found to be invalid, place this warning:
-        # om.add_warning("Invalid data", f"Invalid data found: {key=}; {value=}", info_map)
-        pass
+        return True
+
+    def _validate_num_type_element(self, variable_to_check: Dict[str, Any], var_name: str,
+                                   input_data_value: int) -> bool:
+        """Validates a __pool number element."""
+        info_map = {"class": self.__class__.__name__,
+                    "function": self._validate_num_type_element.__name__,
+                    }
+        is_minimum_check = variable_to_check.get("minimum")
+        is_maximum_check = variable_to_check.get("maximum")
+        if is_minimum_check and is_maximum_check:
+            is_in_range = variable_to_check["minimum"] <= input_data_value <= variable_to_check["maximum"]
+            if not is_in_range:
+                om.add_warning("Value out of range.", f"{var_name=}", info_map)
+                return is_in_range
+        if is_minimum_check:
+            is_in_range = variable_to_check["minimum"] <= input_data_value
+            if not is_in_range:
+                om.add_warning("Value out of range.", f"{var_name=}", info_map)
+                return is_in_range
+        if is_maximum_check:
+            is_in_range = input_data_value <= variable_to_check["maximum"]
+            if not is_in_range:
+                om.add_warning("Value out of range.", f"{var_name=}", info_map)
+                return is_in_range
+        return True
```

Comment:
```suggestion
        minimum_value = variable_to_check.get("minimum")
        maximum_value = variable_to_check.get("maximum")
        is_in_range = True
        if minimum_value is not None and maximum_value is not None:
            is_in_range = minimum_value <= input_data_value <= maximum_value
            warning_string = f"Value not in range [{minimum_value}, {maximum_value}]."
        elif minimum_value is not None:
            is_in_range = minimum_value <= input_data_value
            warning_string = f"Value less than {minimum_value}."
        elif maximum_value is not None:
            is_in_range = input_data_value <= maximum_value
            warning_string = f"Value greater than {maximum_value}."
            
        if not is_in_range:
            om.add_warning(warning_string, f"{var_name=}", info_map)
            
        return is_in_range
```
I think adding the values checked against to the warning would make them a little bit more helpful. Also there is a lot of repeated repeated calls to the OutputManager and `variable_to_check`, by reducing their numbers I think the code becomes more readable.

Reply:
Good call.